File: runtime/src/main/java/com/siimkinks/sqlitemagic/SelectionColumn.java
package com.siimkinks.sqlitemagic;

import android.database.SQLException;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.siimkinks.sqlitemagic.Utils.ValueParser;
import com.siimkinks.sqlitemagic.internal.SimpleArrayMap;
import com.siimkinks.sqlitemagic.internal.StringArraySet;

import java.util.ArrayList;
import java.util.LinkedList;

final class SelectionColumn<
    T, // exact type
    R, // return type (when this column is queried)
    ET, // equivalent type
    P>  // parent table type
    extends NumericColumn<T, R, ET, P> {

  @NonNull
  private final SelectBuilder<?> selectBuilder;
  @Nullable
  private ArrayList<String> parentObservedTables;

  private SelectionColumn(@NonNull Table<P> table, @NonNull String name, boolean allFromTable,
                          @NonNull ValueParser<?> valueParser, boolean nullable,
                          @Nullable String alias, @NonNull String nameInQuery,
                          @NonNull SelectBuilder<?> selectBuilder) {
    super(table, name, allFromTable, valueParser, nullable, alias, nameInQuery);
    this.selectBuilder = selectBuilder;
  }

  @SuppressWarnings("unchecked")
  @NonNull
  static <T> SelectionColumn<T, T, T, ?> from(@NonNull SelectBuilder<?> selectBuilder,
                                              @NonNull String alias) {
    final Select.SingleColumn<?> columnNode = selectBuilder.columnNode;
    if (selectBuilder.columnsNode != null || columnNode == null) {
      throw new SQLException("Only a single result allowed for a SELECT that is part of a column");
    }
    final Table table = selectBuilder.from.table;
    final Column<?, ?, ?, ?> column = columnNode.column;
    return new SelectionColumn<>(table, alias, false, column.valueParser, column.nullable, alias, alias, selectBuilder);
  }

  @Override
  void appendSql(@NonNull StringBuilder sb) {
    sb.append('(');
    selectBuilder.appendCompiledQuery(sb, parentObservedTables);
    sb.append(')');
  }

  @Override
  void appendSql(@NonNull StringBuilder sb, @NonNull SimpleArrayMap<String, LinkedList<String>> systemRenamedTables) {
    sb.append('(');
    selectBuilder.appendCompiledQuery(sb, parentObservedTables);
    sb.append(')');
  }

  @Override
  void addSelectedTables(@NonNull StringArraySet result) {
    // selection column does not add any tables to outer selection as it is returning
    // autonomous column
  }

  @Override
  void addArgs(@NonNull ArrayList<String> args) {
    args.addAll(selectBuilder.args);
  }

  @Override
  void addObservedTables(@NonNull ArrayList<String> tables) {
    // defer observed tables adding until sql building, where we might add missing joins,
    // so there will be more tables to add/observe
    this.parentObservedTables = tables;
  }

  @Override
  int compile(@NonNull SimpleArrayMap<String, Integer> columnPositions,
              @NonNull StringBuilder compiledCols,
              int columnOffset) {
    appendSql(compiledCols);
    appendAliasDeclarationIfNeeded(compiledCols);
    putColumnPosition(columnPositions, name, columnOffset, this);
    return columnOffset + 1;
  }

  @Override
  int compile(@NonNull SimpleArrayMap<String, Integer> columnPositions,
              @NonNull StringBuilder compiledCols,
              @NonNull SimpleArrayMap<String, LinkedList<String>> systemRenamedTables,
              int columnOffset) {
    appendSql(compiledCols, systemRenamedTables);
    appendAliasDeclarationIfNeeded(compiledCols);
    putColumnPosition(columnPositions, name, columnOffset, this);
    return columnOffset + 1;
  }

  @NonNull
  @Override
  public NumericColumn<T, R, ET, P> as(@NonNull String alias) {
    return new SelectionColumn<>(table, name, allFromTable, valueParser, nullable, alias, nameInQuery, selectBuilder);
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/SelectionColumn.kt
package com.siimkinks.sqlitemagic

import android.database.SQLException
import com.siimkinks.sqlitemagic.Utils.ValueParser
import com.siimkinks.sqlitemagic.internal.SimpleArrayMap
import com.siimkinks.sqlitemagic.internal.StringArraySet
import java.util.LinkedList

/**
 * A column used in queries and conditions.
 *
 * @param T Exact type
 * @param R Return type (when this column is queried)
 * @param ET Equivalent type
 * @param P Parent table type
 * @param N Column nullability
 */
internal class SelectionColumn<T, R, ET, P, N> private constructor(
  table: Table<P>,
  name: String,
  allFromTable: Boolean,
  valueParser: ValueParser<*>,
  nullable: Boolean,
  alias: String?,
  nameInQuery: String,
  private val selectBuilder: SelectBuilder<*>
) : NumericColumn<T, R, ET, P, N>(
  table,
  name,
  allFromTable,
  valueParser,
  nullable,
  alias,
  nameInQuery
) {
  private var parentObservedTables: ArrayList<String>? = null
  private var isCompiledToSelection = false

  companion object {
    @JvmStatic
    fun <T, N> from(
      selectBuilder: SelectBuilder<*>,
      alias: String
    ): SelectionColumn<T, T, T, *, N> {
      val columnNode = selectBuilder.columnNode
      if (selectBuilder.columnsNode != null || columnNode == null) {
        throw SQLException("Only a single result allowed for a SELECT that is part of a column")
      }
      val table = checkNotNull(selectBuilder.from).table
      val column = columnNode.column
      return SelectionColumn(
        table = table,
        name = alias,
        allFromTable = false,
        valueParser = column.valueParser,
        nullable = column.nullable,
        alias = alias,
        nameInQuery = alias,
        selectBuilder = selectBuilder
      )
    }
  }

  override fun appendSql(sb: StringBuilder) {
    if (isCompiledToSelection) {
      sb.append(getAppendableAlias())
      return
    }
    sb.append('(')
    selectBuilder.appendCompiledQuery(sb, parentObservedTables)
    sb.append(')')
  }

  override fun appendSql(
    sb: StringBuilder,
    systemRenamedTables: SimpleArrayMap<String, LinkedList<String>>
  ) = appendSql(sb)

  override fun addSelectedTables(result: StringArraySet) {
    // selection column does not add any tables to outer selection as it is returning
    // an autonomous column
  }

  override fun addArgs(args: ArrayList<String>) {
    args.addAll(selectBuilder.args)
  }

  override fun addObservedTables(tables: ArrayList<String>) {
    // defer observed tables adding until SQL building, where we might add missing joins,
    // so there will be more tables to add/observe.
    parentObservedTables = tables
  }

  override fun compile(
    columnPositions: SimpleArrayMap<String, Int>,
    compiledCols: StringBuilder,
    columnOffset: Int
  ): Int {
    appendSql(compiledCols)
    appendAliasDeclarationIfNeeded(compiledCols)
    putColumnPosition(
      columnPositions = columnPositions,
      columnId = name,
      pos = columnOffset,
      column = this
    )
    isCompiledToSelection = true
    return columnOffset + 1
  }

  override fun compile(
    columnPositions: SimpleArrayMap<String, Int>,
    compiledCols: StringBuilder,
    systemRenamedTables: SimpleArrayMap<String, LinkedList<String>>,
    columnOffset: Int
  ): Int {
    appendSql(compiledCols, systemRenamedTables)
    appendAliasDeclarationIfNeeded(compiledCols)
    putColumnPosition(
      columnPositions = columnPositions,
      columnId = name,
      pos = columnOffset,
      column = this
    )
    isCompiledToSelection = true
    return columnOffset + 1
  }

  override fun `as`(alias: String): NumericColumn<T, R, ET, P, N> = SelectionColumn(
    table = table,
    name = name,
    allFromTable = allFromTable,
    valueParser = valueParser,
    nullable = nullable,
    alias = alias,
    nameInQuery = nameInQuery,
    selectBuilder = selectBuilder
  )

  override fun <NewTableType> inTable(table: Table<NewTableType>): NumericColumn<T, R, ET, NewTableType, N> =
    SelectionColumn(
      table = table,
      name = name,
      allFromTable = allFromTable,
      valueParser = valueParser,
      nullable = nullable,
      alias = alias,
      nameInQuery = nameInQuery,
      selectBuilder = selectBuilder
    )
}
